core/src/main/scala/kafka/server/KafkaConfig.scala

```diff
@@ -2018,12 +2015,104 @@ class KafkaConfig private(doLog: Boolean, val props: java.util.Map[_, _], dynami
+        s"If ${KafkaConfig.ProcessRolesProp} contains just the 'broker' role, the node id $nodeId must not be included in the set of voters ${KafkaConfig.QuorumVotersProp}=${voterAddressSpecsByNodeId.asScala.keySet.toSet}")
+      // controller.listener.names must be non-empty...
+      require(controllerListenerNames.exists(_.nonEmpty),
+        s"${KafkaConfig.ControllerListenerNamesProp} must contain at least one value when running KRaft with just the broker role")
+      // controller.listener.names are forbidden in listeners...
+      require(controllerListeners.isEmpty,
+        s"${KafkaConfig.ControllerListenerNamesProp} must not contain a value appearing in the '${KafkaConfig.ListenersProp}' configuration when running KRaft with just the broker role")
+      // controller.listener.names must all appear in listener.security.protocol.map
+      controllerListenerNames.filter(_.nonEmpty).foreach { name =>
+        val listenerName = ListenerName.normalised(name)
+        if (!listenerSecurityProtocolMap.contains(listenerName)) {
+          throw new ConfigException(s"Controller listener with name ${listenerName.value} defined in " +
+            s"${KafkaConfig.ControllerListenerNamesProp} not found in ${KafkaConfig.ListenerSecurityProtocolMapProp}.")
+        }
+      }
+      // warn that only the first controller listener is used if there is more than one
+      if (controllerListenerNames.size > 1) {
+        warn(s"${KafkaConfig.ControllerListenerNamesProp} has multiple entries; only the first will be used since ${KafkaConfig.ProcessRolesProp}=broker: ${controllerListenerNames.asJava.toString}")
+      }
+      validateAdvertisedListenersNonEmptyForBroker()
+    } else if (processRoles == Set(ControllerRole)) {
+      // KRaft controller-only
+      validateCanParseControllerQuorumVotersForKRaft()
+      validateControlPlaneListenerEmptyForKRaft()
+      // advertised listeners must be empty when not also running the broker role
+      require(advertisedListeners.isEmpty,
+        s"$sourceOfAdvertisedListeners must only contain KRaft controller listeners from ${KafkaConfig.ControllerListenerNamesProp} when ${KafkaConfig.ProcessRolesProp}=controller")
+      validateControllerQuorumVotersMustContainNodeIDForKRaftController()
+      validateControllerListenerExistsForKRaftController()
+      validateControllerListenerNamesMustAppearInListenersForKRaftController()
+    } else if (processRoles == Set(BrokerRole, ControllerRole)) {
+      // KRaft colocated broker and controller
+      validateCanParseControllerQuorumVotersForKRaft()
+      validateControlPlaneListenerEmptyForKRaft()
+      validateAdvertisedListenersDoesNotContainControllerListenersForKRaftBroker()
+      validateControllerQuorumVotersMustContainNodeIDForKRaftController()
+      validateControllerListenerExistsForKRaftController()
+      validateControllerListenerNamesMustAppearInListenersForKRaftController()
+      validateAdvertisedListenersNonEmptyForBroker()
```

Comment:
Correct me if I'm wrong, but I think the validations here imply that mixed nodes _must_ specify `advertised.listeners`. If the user didn't provide `advertised.listeners`, then we would take the effective advertised listeners from `listeners`, which would have to include the controller listener. In other words, this would be invalid:
```
process.roles=broker,controller
controller.listener.names=CONTROLLER
listeners=DEFAULT://:9092,CONTROLLER://:9093
```

Instead, we would have to do this:
```
process.roles=broker,controller
controller.listener.names=CONTROLLER
listeners=DEFAULT://:9092,CONTROLLER://:9093
advertised.listeners=DEFAULT://:9092
```

Does that seem right? An alternative approach we could use here would be to automatically exclude the controller listeners from the advertised listeners when they are specified by `listeners`. The configuration itself gives us enough information to do the right thing.

By the way, would it make sense to rename `advertisedListeners` to `effectiveAdvertisedListeners` to make it clearer that the source might not be `advertised.listeners`?

Reply:
The config you presented as being illegal is actually acceptable assuming we also set `inter.broker.listener.name` and `listener.secuerity.protocol.map`:

```
process.roles=broker,controller
controller.listener.names=CONTROLLER
listeners=DEFAULT://:9092,CONTROLLER://:9093
# need to add these two configs to make it work
inter.broker.listener.name=DEFAULT
listener.security.protocol.map=DEFAULT:PLAINTEXT,CONTROLLER:PLAINTEXT
```

We could make it simpler if we take advantage of various default values, like this:

```
process.roles=broker,controller
controller.listener.names=CONTROLLER
listeners=PLAINTEXT://:9092,CONTROLLER://:9093
# inter.broker.listener.name and litener.security.protocol.map default values will now allow it to work
```

> An alternative approach we could use here would be to automatically exclude the controller listeners from the advertised listeners when they are specified by listeners

Yes, this is what we do:
```
  def advertisedListeners: Seq[EndPoint] = {
...
      listeners.filterNot(l => controllerListenerNames.contains(l.listenerName.value()))
  }
```

> By the way, would it make sense to rename advertisedListeners to effectiveAdvertisedListeners to make it clearer that the source might not be advertised.listeners?

Yes!  I think that makes perfect sense.

Reply:
Ah, interesting. I missed the filtering of `listeners`.